recipes/LibriSpeech/SSL/train.py

```diff
@@ -0,0 +1,518 @@
+    then collate them but at that time one doesn't know the length of the
+    shortest sample in the batch (which determines the number of masked frames)
+    so it's better this way.
+    This function also has the feature to split samples longer than `max_dur` into
+    smaller chunks, which helps reduce the memory usage of attention based systems.
+
+    Arguments
+    ---------
+    samples_lst : list
+        List of samples returned by the audio_pipeline.
+    get_out_len_fn : function
+        Function that calculates length of sample after it passes through feature extractor.
+    mask_prob : float
+        Approximate percentage of frames to mask.
+    mask_length : int
+        Number of contiguous frames that will be masked.
+    max_dur : float
+        Samples longer than this will be split so they're shorter.
+    Returns
+    -------
+    wavs_padded : torch.Tensor, shape (B, T)
+        Audio arrays with right-sided padding.
+    wav_lens : torch.Tensor, shape (B,)
+        For each sample the percentage of the array that is not padding.
+    mask : torch.Tensor, shape (B, T)
+        Boolean mask to mask frames.
+    """
+    wav_lst, latent_length_lst = [], []
+    # If the sample is longer than `max_dur` create several shorter samples from it.
+    # Using max of all samples because if we considered the splitting for each sample separately
+    # then it could happen that some samples would be just below `max_dur` and others just above,
+    # and splitting those above would result in an unnecessarily large amount of padding.
+    max_sample_dur = max(
+        sample["sig"].size(0) / 16000 for sample in samples_lst
+    )
+    will_break = max_sample_dur > max_dur
+    splits = math.ceil(max_sample_dur / max_dur)
+    for sample in samples_lst:
+        sig = sample["sig"]
+        if will_break:
+            for i in range(splits):
+                sample_len_split = math.ceil(sample["sig"].size(0) / splits)
+                wav_lst.append(
+                    sample["sig"][
+                        i * sample_len_split : (i + 1) * sample_len_split
+                    ]
+                )
+                latent_length = get_out_len_fn(
+                    torch.as_tensor(wav_lst[-1].size(0))
+                ).item()
+                latent_length_lst.append(latent_length)
+        else:
+            wav_lst.append(sig)
+            latent_length = get_out_len_fn(torch.as_tensor(sig.size(-1)))
+            latent_length_lst.append(latent_length.item())
+    bs = len(wav_lst)
+    wavs_padded, wav_lens = batch_pad_right(wav_lst)
+
+    batch_time_len = max(latent_length_lst)
+    mask = compute_mask(
+        (bs, batch_time_len,), latent_length_lst, mask_prob, mask_length
+    )
+    return (
+        torch.as_tensor(wavs_padded),
+        torch.as_tensor(wav_lens),
+        torch.as_tensor(mask, dtype=torch.bool),
+    )
+
+
+def _setup_wandb(hparams):
```

Comment:
Did you try this with multinode training ? If not, we should, because this often break when multinode training is used.

Reply:
I always do multinode training